python/src/uagents/communication.py

```diff
@@ -0,0 +1,268 @@
+from uagents.models import Model
+from uagents.resolver import GlobalResolver, Resolver, parse_identifier
+
+LOGGER = get_logger("dispenser", logging.DEBUG)
+
+
+class DeliveryStatus(str, Enum):
+    """Delivery status of a message."""
+
+    SENT = "sent"
+    DELIVERED = "delivered"
+    FAILED = "failed"
+    QUEUED = "queued"
+
+
+@dataclass
+class MsgDigest:
+    """
+    Represents a message digest containing a message and its schema digest.
+
+    Attributes:
+        message (Any): The message content.
+        schema_digest (str): The schema digest of the message.
+    """
+
+    message: Any
+    schema_digest: str
+
+
+@dataclass
+class MsgStatus:
+    """
+    Represents the status of a sent message.
+
+    Attributes:
+        status (str): The delivery status of the message {'sent', 'delivered', 'failed'}.
+        detail (str): The details of the message delivery.
+        destination (str): The destination address of the message.
+        endpoint (str): The endpoint the message was sent to.
+    """
+
+    status: DeliveryStatus
+    detail: str
+    destination: str
+    endpoint: str
+
+
+class Dispenser:
+    """
+    Dispenses messages externally.
+    """
+
+    def __init__(self):
+        self._envelopes: List[Tuple[Envelope, bool]] = []
+        self._timeout: float = DEFAULT_DISPENSER_INTERVAL
+        self._resolver: Resolver = GlobalResolver()
+
+    def configure(self, resolver: Resolver, timeout: Optional[float] = None):
+        if resolver:
+            self._resolver = resolver
+        if timeout:
+            self._timeout = timeout
+
+    def add_envelope(self, envelope: Envelope, sync: bool = False):
+        self._envelopes.append((envelope, sync))
+
+    async def run(self):
+        while True:
+            for env, sync in self._envelopes:
+                try:
+                    await send_exchange_envelope(
+                        envelope=env,
+                        resolver=self._resolver,
+                        sync=sync,
+                    )  # how do we handle the MsgStatus response?
+                    self._envelopes.remove((env, sync))
+                except ValueError:
+                    continue
+            await asyncio.sleep(DEFAULT_DISPENSER_INTERVAL)
+
+
+async def dispatch_local_message(
+    sender: str,
+    destination: str,
+    schema_digest: str,
+    message: JsonStr,
+    session_id: uuid.UUID,
+) -> MsgStatus:
+    """Process a message locally."""
+    await dispatcher.dispatch(
+        sender=sender,
+        destination=destination,
+        schema_digest=schema_digest,
+        message=message,
+        session=session_id,
+    )
+    return MsgStatus(
+        status=DeliveryStatus.DELIVERED,
+        detail="Message dispatched locally",
+        destination=destination,
+        endpoint="",
+    )
+
+
+async def send_exchange_envelope(
+    envelope: Envelope,
+    resolver: Optional[Resolver] = None,
+    sync: bool = False,
+) -> Union[MsgStatus, Envelope]:
```

Comment:
The destination of the envelope actually depends on the output of the resolver, so the envelope may need to be created inside this function.

Reply:
I wasn't aware of that and unfortunately the naming of `destination` and `destination_address` was a bit misleading. In all of my tests those where identical but I have to admit, I didn't try the ANS.
But this means that the resolver logic needs to be put back into the context (which I'm fine with) as the goal here should be to have the `context.send()` method create the `Envelope` and the actual sending happens elsewhere.

Reply:
That works, but in that case, we might want to create a standalone wrapper for `send_exchange_envelope` that constructs the envelope similar to the way this function does: https://github.com/fetchai/uAgents/blob/main/python/src/uagents/context.py#L534-L548.